src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.Async.cs

```diff
@@ -0,0 +1,203 @@
+        bool result = TryReadBlockCore(blockContents.AsSpan(), bytesRead, out Zip64EndOfCentralDirectoryLocator zip64EOCDLocator);
+        return (result, zip64EOCDLocator);
+    }
+
+    public static ValueTask WriteBlockAsync(Stream stream, long zip64EOCDRecordStart, CancellationToken cancellationToken)
+    {
+        byte[] blockContents = new byte[TotalSize];
+        WriteBlockCore(blockContents.AsSpan(), zip64EOCDRecordStart);
+        return stream.WriteAsync(blockContents, cancellationToken);
+    }
+}
+
+internal sealed partial class Zip64EndOfCentralDirectoryRecord
+{
+    public static async Task<(bool, Zip64EndOfCentralDirectoryRecord)> TryReadBlockAsync(Stream stream, CancellationToken cancellationToken)
+    {
+        byte[] blockContents = new byte[BlockConstantSectionSize];
+        int bytesRead = await stream.ReadAsync(blockContents, cancellationToken).ConfigureAwait(false);
+        bool result = TryReadBlockCore(blockContents, bytesRead, out Zip64EndOfCentralDirectoryRecord zip64EOCDRecord);
+        return (result, zip64EOCDRecord);
+    }
+
+    public static async Task WriteBlockAsync(Stream stream, long numberOfEntries, long startOfCentralDirectory, long sizeOfCentralDirectory, CancellationToken cancellationToken)
+    {
+        byte[] blockContents = new byte[BlockConstantSectionSize];
+        WriteBlockCore(blockContents.AsSpan(), numberOfEntries, startOfCentralDirectory, sizeOfCentralDirectory);
+        // write Zip 64 EOCD record
+        await stream.WriteAsync(blockContents, cancellationToken).ConfigureAwait(false);
+    }
+}
+
+internal readonly partial struct ZipLocalFileHeader
+{
+    public static async Task<List<ZipGenericExtraField>> GetExtraFieldsAsync(Stream stream, CancellationToken cancellationToken)
+    {
+        // assumes that TrySkipBlock has already been called, so we don't have to validate twice
+
+        List<ZipGenericExtraField> result;
+        int relativeFilenameLengthLocation = FieldLocations.FilenameLength - FieldLocations.FilenameLength;
+        int relativeExtraFieldLengthLocation = FieldLocations.ExtraFieldLength - FieldLocations.FilenameLength;
+        byte[] fixedHeaderBuffer = new byte[FieldLengths.FilenameLength + FieldLengths.ExtraFieldLength];
+
+        stream.Seek(FieldLocations.FilenameLength, SeekOrigin.Current);
+        await stream.ReadExactlyAsync(fixedHeaderBuffer, cancellationToken).ConfigureAwait(false);
+
+        ushort filenameLength = BinaryPrimitives.ReadUInt16LittleEndian(fixedHeaderBuffer.AsSpan(relativeFilenameLengthLocation));
+        ushort extraFieldLength = BinaryPrimitives.ReadUInt16LittleEndian(fixedHeaderBuffer.AsSpan(relativeExtraFieldLengthLocation));
+        byte[] extraFieldBuffer = ArrayPool<byte>.Shared.Rent(extraFieldLength);
+
+        try
+        {
+            stream.Seek(filenameLength, SeekOrigin.Current);
+            await stream.ReadExactlyAsync(extraFieldBuffer, cancellationToken).ConfigureAwait(false);
```

Comment:
`extraFieldBuffer` is rented from an ArrayPool, and might thus be larger than `extraFieldLength`.